### PR TITLE
Chemistry JSON dump tool and companion GitHub Action

### PR DESCRIPTION
**Pull request by weaversam8**
## About the PR 

This is a continuation of @moonheart08's work in #5891. The main changes are:

- Updated JSON generation to use `System.Text.Json` instead of `Newtonsoft.Json`
- Implemented a partial fix for the crash mentioned in #5891.
  - Running the generator still crashes with a nonzero error code, but if/when space-wizards/RobustToolbox#2417 is merged, running the generator will now successfully shutdown the server after file generation.
- I also created a GitHub Actions Workflow to automatically upload updated versions of the autogenerated JSON files to the SS14 MediaWiki instance. More info below:

### How to use the generator

As in #5891, simply run the server with `--cvar autogen.destination_file=prototypes.json` and JSON files will be generated in `./bin/Content.Server/data` and will be will be:
- `chem_prototypes.json`
- `react_prototypes.json`

These files support @moonheart08's auto-generated Chemistry documentation on the SS14 wiki, which you can check out in more detail [here](https://wiki.spacestation14.io/w/index.php?title=Chemistry&action=edit).

### How to use the new GH Actions Workflow

The new GitHub actions workflow will automatically build the server and run it to generate the above JSON files. It will then attempt to upload those files to the MediaWiki instance using [jtmullen/mediawiki-edit-action](https://github.com/jtmullen/mediawiki-edit-action). In order to do that, the following [secrets](https://docs.github.com/en/actions/security-guides/encrypted-secrets#creating-encrypted-secrets-for-a-repository) must be configured on the repository:

- `WIKI_PAGE_ROOT` - the root page where JSON files will be nested under.
  - To enable this Actions bot for the existing bot pages, this value should be set to `User:MoMMI`, but configuring it with another bot user is equally possible.
- `WIKI_BOT_USER` - the MediaWiki user account to authenticate with
- `WIKI_BOT_PASS` - the password for said account

---

Totally open to feedback on this PR, by the way. Not sure whether the approach I took is sufficient or needs to be more robust.

### CODE REVIEW

**Comment by weaversam8** [comment]
Update: not yet ready to merge due a few issues remaining with serialization after the `System.Text.Json` conversion. [(See Discord for more context.)](https://discord.com/channels/310555209753690112/311537926376783886/930684045388558377). Will update with a fix in the next few days.

**Comment by weaversam8** [comment]
Okay, this should be ready now.

The reason why this took longer than expected is that `System.Text.Json` handles serialization of polymorphic objects different than `Newtonsoft.Json`. In order to ensure we didn't regress in functionality for our serializations of `ReagentEntry`, `ReagentEffect`, `ReagentEffectCondition`, `ReagentPrototype`, and `DamageSpecifier` classes, I had to implement a [custom converter](https://docs.microsoft.com/en-us/dotnet/standard/serialization/system-text-json-converters-how-to?pivots=dotnet-6-0) which could handle serializing polymorphic objects correctly.

This new class is called `UniversalJsonConverter` and is in the (new) `Content.Shared.Converters` namespace. It will be useful to anyone attempting to serialize polymorphic objects in the future.
- For examples of its usage, see one of the classes mentioned above (adding the custom converter is pretty easy.)
- For an explanation of why this class was necessary, see [the "Polymorphic Serialization" section](https://docs.microsoft.com/en-us/dotnet/standard/serialization/system-text-json-migrate-from-newtonsoft-how-to?pivots=dotnet-6-0#polymorphic-serialization) of the `System.Text.Json` docs.
- For an explanation of how this class works, check out [the source itself](https://github.com/space-wizards/space-station-14/blob/c9b1c0ac65e825a4fa685e3c23b737ab88dcea9b/Content.Shared/Converters/UniversalJsonConverter.cs), which should be relatively well-documented.

### Before merging

1. @moonheart08, please confirm that the JSON files generated by the new code are semantically compatible with the files currently in place on the wiki. You may find [JSON Diff - The semantic JSON compare tool](http://jsondiff.com) helpful.
    - Old files, under MoMMI bot:
      - [`react_prototypes.json`](https://wiki.spacestation14.io/w/index.php?title=User:MoMMI/react_prototypes.json)
      - [`chem_prototypes.json`](https://wiki.spacestation14.io/w/index.php?title=User:MoMMI/chem_prototypes.json)
    - New files, under Weaver/NexBot:
      - [`react_prototypes.json`](https://wiki.spacestation14.io/w/index.php?title=User:Weaver/NexBot/react_prototypes.json)
      - [`chem_prototypes.json`](https://wiki.spacestation14.io/w/index.php?title=User:Weaver/NexBot/chem_prototypes.json)
2. @mirrorcult or another person with permissions, please add the following environment variables (mentioned above) to [the Secrets section of the repository settings](https://github.com/space-wizards/space-station-14/settings/secrets/actions), so the new "Build and publish JSON for the wiki" action can run successfully on future commits:
    - `WIKI_PAGE_ROOT` - the root page where JSON files will be nested under.
      - To enable this Actions bot for the existing bot pages, this value should be set to `User:MoMMI`, but configuring it with another bot user is equally possible.
    - `WIKI_BOT_USER` - the MediaWiki user account to authenticate with
    - `WIKI_BOT_PASS` - the password for said account

**Comment by moonheart08** [comment]
Confirming this renders correctly on the wiki, so this is ready-to-use.

**Comment by mirrorcult** [comment]
doesn't look like I have perms to edit the action secrets

File: .github/workflows/update-wiki.yml
Comment (mirrorcult) [inline .github/workflows/update-wiki.yml:13]: imo this should probably only push on chem-related changes. but maybe that is hard to specify

File: Content.Server/Entry/EntryPoint.cs
Comment (mirrorcult) [inline Content.Server/Entry/EntryPoint.cs:113]: this is very ugly but i do not know a better method

**Review by mirrorcult — COMMENTED** [review]
it would be so extremely nice if we could forgo json entirely and just use serv3 since we already have datafields and stuff for it + it can do polymorphic se/deser. but that does not need to be done here (as i imagine that also means writing a yaml parser for mediawiki which uhhhhh)

**Comment by moonheart08** [comment]
> it would be so extremely nice if we could forgo json entirely and just use serv3 since we already have datafields and stuff for it + it can do polymorphic se/deser. but that does not need to be done here (as i imagine that also means writing a yaml parser for mediawiki which uhhhhh)

Yea using serv3 means teaching mediawiki how YAML works.
Maybe a future stretch goal, but for now it's out of the question; JSON loading comes built-in.

File: .github/workflows/update-wiki.yml
Comment (moonheart08) [reply .github/workflows/update-wiki.yml:13]: At the very least it doesn't cost us anything if it pushes too frequently.

File: Content.Server/Entry/EntryPoint.cs
Comment (weaversam8) [reply Content.Server/Entry/EntryPoint.cs:113]: open to suggestions on a better place to put this. I don't think it should block this PR, but since I plan to add more auto-generated wiki stuff later, we'll get a chance to revisit this.

File: .github/workflows/update-wiki.yml
Comment (weaversam8) [reply .github/workflows/update-wiki.yml]: This has been updated to a narrower list of globs, resolving.

**Review by moonheart08 — APPROVED** [review]
Stuff that isn't me-code looks good.

**Review by mirrorcult — APPROVED** [review]

**Pull request merged**
